Rename `torch._C._TensorBase` to `TensorBase`

fleonce opened the pull request:
I have gone ahead and implemented the renaming of the type `torch._C._TensorBase` to a non-private class name `TensorBase`.
The changes also include leaving `torch._C._TensorBase` as an alias to the new type: https://github.com/pytorch/pytorch/blob/70458768fbefbb1dbda87bf645e14d79f44f67ef/torch/csrc/autograd/python_variable.cpp#L2196-L2197 both in the c++ code and in the corresponding `__init__.pyi.in` file:
https://github.com/pytorch/pytorch/blob/70458768fbefbb1dbda87bf645e14d79f44f67ef/torch/_C/__init__.pyi.in#L1522

Fixes #109438

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ezyang @malfet

fleonce committed:
Make `torch._C._TensorBase` a non-private class, update references to `TensorBase` (#109438)

Summary:

- Renamed `tp_name` of `THPVariableType` to `torch._C.TensorBase` in `torch/csrc/autograd/python_variable.cpp`
  Change the name of the underlying C bound type for `Tensor` to a non-private name: `TensorBase`

- Added new field `TensorBase` with reference to the new `TensorBase` type
  to `torch._C`, keep the old reference to `_TensorBase` to not break backwards
  compatibility.

- Update `torch/_C/__init__.pyi.in` to include the new `TensorBase` type as
  well as an alias `_TensorBase` to stay consistent with the cpp changes

fleonce committed:
Rename references of `_TensorBase` to `TensorBase` (#109438)

ezyang commented:
Did you confirm this fixed autocomplete in the IDE?

fleonce commented:
I can confirm this works on PyCharm 2023.2 (Build #PY-232.8660.197):

Both method type hints and method suggestions in general work now, with both originating from TensorBase, as intended:
![image](https://github.com/pytorch/pytorch/assets/8986525/f0a13fea-0b3d-4ce1-a93d-8749390d25b3)
![image](https://github.com/pytorch/pytorch/assets/8986525/f3a078e7-0d34-4b6e-b14e-fe1d9e0bf2fc) 
It also works now to open the parent class of the Python `torch.Tensor` class (Inside the Editor).

fleonce commented:
The tests seem to fail because `TensorBase` is not in the `torch_C_allowlist_superset` in https://github.com/pytorch/pytorch/blob/main/test/test_public_bindings.py#L28

ezyang commented:
Oh, you need to stop TensorBase from being copied into torch.

That being said, it looks like this should happen already?

```
for name in dir(_C):
    if name[0] != '_' and not name.endswith('Base'):
        __all__.append(name)
        obj = getattr(_C, name)
        if (isinstance(obj, Callable) or inspect.isclass(obj)):  # type: ignore[arg-type]
            if (obj.__module__ != 'torch'):
                # TODO: fix their module from C++ side
                if name not in ['DisableTorchFunctionSubclass', 'DisableTorchFunction', 'Generator']:
                    obj.__module__ = 'torch'
```

if you can confirm with your patch that `torch.TensorBase` doesn't exist you can just whitelist it for the test.

fleonce commented:
I think it is not happening that way, but rather due to the import statements in `torch/__init__.py`:
```python3
from torch._C import *
``` 
```python3
>>> import torch
>>> torch.TensorBase
<class 'torch._C.TensorBase'>
>>> "TensorBase" in torch.__all__
False
```
The same actually happens with `torch._C.StorageBase`, which is not in `__all__` either:
```python3
>>> torch.StorageBase
<class 'torch._C.StorageBase'>
>>> "StorageBase" in torch.__all__
False
```
But `StorageBase` is whitelisted in the Test: https://github.com/pytorch/pytorch/blob/main/test/test_public_bindings.py#L61

fleonce commented:
When constructing `__all__`, I could `delattr` class names that end with `Base`? Not sure how to approach this correctly

fleonce commented:
The check also does not check if f.e. `torch.TensorBase` is actually present but assumes it for all type names not starting with an underscore: https://github.com/pytorch/pytorch/blob/main/test/test_public_bindings.py#L221
`torch_C_bindings = {elem for elem in dir(torch._C) if not elem.startswith("_")}`, so even if I do
`delattr(sys.modules[__name__], "TensorBase")` in `torch/__init__.py`, the check will still fail I think

ezyang commented:
I think delattr is the way to go

fleonce commented:
I managed to remove TensorBase from torch:
```python3
>>> import torch
>>> torch.TensorBase
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/moritz_hennen/projects/pytorch/torch/__init__.py", line 1851, in __getattr__
    raise AttributeError(f"module '{__name__}' has no attribute '{name}'")
AttributeError: module 'torch' has no attribute 'TensorBase'
```
but the test still gives the same error:
```
AssertionError: False is not true : torch._C had bindings that are not present in the allowlist:
{'TensorBase'}

To execute this test, run the following from the base repo dir:
     python test/test_public_bindings.py -k test_no_new_bindings
```

fleonce committed:
Prevent `torch._C.TensorBase` from being copied into `torch.` (#109940)

`TensorBase` gets copied into the main torch package, due
to the import statements in `torch/__init__.py`. Following the same
logic of `torch.__all__`, use `delattr` to remove all types from
`torch._C` that are both public and end with `Base`

fleonce committed:
Add `TensorBase` to `torch_C_allowlist_superset` (#109940)

Even though `torch._C.TensorBase` is not copied into `torch.`,
`test/test_public_bindings.py` will still fail because it
assumes all public types from `torch._C` to be present as `torch.XXX`.

To prevent the test from failing, `TensorBase` has to be included in the allowlist.

fleonce commented:
I now have implemented the following rule in `torch/__init__.py`:

```python3
    elif name[0] != '_' and name.endswith('Base'):
        # issue 109438 / pr 109940. Prevent TensorBase from being copied into torch.
        delattr(sys.modules[__name__], name)
```
this also filters `torch._C.StorageBase`, which I think is intended anyway

For the failing test I have added `TensorBase` as an exclusion, maybe a better way to test would be checking whether or not the types from `torch._C` are also actually present as an attribute as `torch.XXX` before comparing the allowed list with the present values

ezyang commented on torch/__init__.py:
```diff
@@ -472,6 +472,9 @@ def sym_min(a, b):
                     obj.__module__ = 'torch'
+    elif name[0] != '_' and name.endswith('Base'):
+        # issue 109438 / pr 109940. Prevent TensorBase from being copied into torch.
+        delattr(sys.modules[__name__], name)
```
Can you just match TensorBase explicitly, I want to avoid hitting any other identifiers accidentally.

ezyang commented on test/test_public_bindings.py:
```diff
@@ -199,6 +199,7 @@ def test_no_new_bindings(self):
             "SUM",
             "SymFloat",
             "SymInt",
+            "TensorBase",
```
It's annoying that you have to put this here, because you're actually explicitly deleting it. Maybe put this in a separate list that gets unioned together before the check?

ezyang commented:
This is looking good, just one small nit.

fleonce committed:
Explicitly filter `TensorBase` from the torch package (#109940)

Instead of filtering for types from _C that end with 'Base', filter just 'TensorBase'.

Intruduced another set of explicitly removed classes for otherwise `torch.XXX` types
that come from `torch._C` and are not private for the public bindings test.

fleonce commented:
Alright, I have changed both files following your review comments @ezyang

ezyang commented:
@pytorchbot merge
